### PR TITLE
remove non printable unicode chars in category name

### PR DESCRIPTION
**Pull request by WMDE-Fisch**
I did not find any info about allowed or filtered non printable chars in category names / page titles for Mediawiki so I implemented a custom one.

see https://phabricator.wikimedia.org/T124161
and https://phabricator.wikimedia.org/T123897

### CODE REVIEW

**Comment by wmde-manicki** [comment]
Looks good as far as I can tell. This regular expression is of course awesome and figuring all things it is removing might be quite some work...
Would be nice to have some more tests case, so it is directly shown we're not dropping some characters that are actually relevant. How about cases like:
- digits and some punctuation (e.g. parentheses, comma),
- umlauts and possible some other Latin characters with diacritics like "è", 
- some text in non-Latin script, I'd say Cyrillic, some Chinese and maybe something which is written right-to-left (Arabic?) should be just fine.

Adding some other non-printable characters to test might be also cool but I cannot think of what we could easily add :)

**Comment by wmde-manicki** [comment]
Nyan-Кошки !
Thanks, great!

File: DeepCat.js [hunk 1]
```diff
@@ -583,6 +583,17 @@
 	}
 
 	/**
+	 * Hash function for generating hint box cookie token.
```
Comment (thiemowmde) [inline DeepCat.js:586]: How is this a "hash function"?

File: DeepCat.js [hunk 2]
```diff
@@ -583,6 +583,17 @@
...
...206F\u2072\u2073\u208F\u209D-\u209F\u20BB-\u20CF\u20F1-\u20FF\u218A-\u218F\u23F4-\u23FF\u2427-\u243F\u244B-\u245F\u2700\u2B4D-\u2B4F\u2B5A-\u2BFF\u2C2F\u2C5F\u2CF4-\u2CF8\u2D26\u2D28-\u2D2C\u2D2E\u2D2F\u2D68-\u2D6E\u2D71-\u2D7E\u2D97-\u2D9F\u2DA7\u2DAF\u2DB7\u2DBF\u2DC7\u2DCF\u2DD7\u2DDF\u2E3C-\u2E7F\u2E9A\u2EF4-\u2EFF\u2FD6-\u2FEF\u2FFC-\u2FFF\u3040\u3097\u3098\u3100-\u3104\u312E-\u3130\u318F\u31BB-\u31BF\u31E4-\u31EF\u321F\u32FF\u4DB6-\u4DBF\u9FCD-\u9FFF\uA48D-\uA48F\uA4C7-\uA4CF\uA62C-\uA63F\uA698-\uA69E\uA6F8-\uA6FF\uA78F\uA794-\uA79F\uA7AB-\uA7F7\uA82C-\uA82F\uA83A-\uA83F\uA878-\uA87F\uA8C5-\uA8CD\uA8DA-\uA8DF\uA8FC-\uA8FF\uA954-\uA95E\uA97D-\uA97F\uA9CE\uA9DA-\uA9DD\uA9E0-\uA9FF\uAA37-\uAA3F\uAA4E\uAA4F\uAA5A\uAA5B\uAA7C-\uAA7F\uAAC3-\uAADA\uAAF7-\uAB00\uAB07\uAB08\uAB0F\uAB10\uAB17-\uAB1F\uAB27\uAB2F-\uABBF\uABEE\uABEF\uABFA-\uABFF\uD7A4-\uD7AF\uD7C7-\uD7CA\uD7FC-\uF8FF\uFA6E\uFA6F\uFADA-\uFAFF\uFB07-\uFB12\uFB18-\uFB1C\uFB37\uFB3D\uFB3F\uFB42\uFB45\uFBC2-\uFBD2\uFD40-\uFD4F\uFD90\uFD91\uFDC8-\uFDEF\uFDFE\uFDFF\uFE1A-\uFE1F\uFE27-\uFE2F\uFE53\uFE67\uFE6C-\uFE6F\uFE75\uFEFD-\uFF00\uFFBF-\uFFC1\uFFC8\uFFC9\uFFD0\uFFD1\uFFD8\uFFD9\uFFDD-\uFFDF\uFFE7\uFFEF-\uFFFB\uFFFE\uFFFF]/g;
```
Comment (thiemowmde) [inline DeepCat.js:592]: I would not have approved this. Almost all of this is not needed, except for very few characters. Why not use a small regex that removes exactly the character you stumbled upon? Which is the LEFT-TO-RIGHT MARK (`\u200E`), typically the top reason for such issues.

~~But even this should not be removed everywhere, only from the end of the string. Some wikis really have category names with this character in the middle of the string, for a good reason.~~

Or to put this the other way around: The rule set to normalize page names must be the same in the code that builds the index and in the code that queries the index. Removing all LTR is probably ok then.

But even this code should not be here but in the backend.

**Comment by wmde-manicki** [comment]
Thank you @thiemowmde for commenting! We will be re-fixing that.

> But even this should not be removed everywhere, only from the end of the string. Some wikis really have category names with this character in the middle of the string, for a good reason.

So true. I totally forgot about such cases.

> Or to put this the other way around: The rule set to normalize page names must be the same in the code that builds the index and in the code that queries the index. Removing all LTR is probably ok then.

That is also true. I'd assume backend is not doing any sort of such cleaning as it just assumes it gets an existing category name. So maybe we'd rather try to remove stuff which is not belonging to the category name but just got inserted by mistake (by copy&paste etc). LTR mark will be included, maybe some weird stuff like tab or backspace too. But you're right it shouldn't be dropping LTR always, and this regexp i probably doing far more bad things!

File: DeepCat.js (hunk 1 above)
Comment (wmde-manicki) [inline DeepCat.js:586]: haha, good point!

**Comment by thiemowmde** [comment]
I found the relevant lines in core:

https://github.com/wikimedia/mediawiki/blob/master/includes/title/MediaWikiTitleCodec.php#L220

This means \u200E (0xE2 0x80 0x8E), \u200F (0xE2 0x80 0x8F) and \u202A-\u202E (0xE2 0x80 0xAA to 0xE2 0x80 0xAE) are indeed removed, everywhere.

I confused this with the zero width space (\u200B), which is actually used in category names.

File: DeepCat.js (hunk 1 above)
Comment (WMDE-Fisch) [inline DeepCat.js:586]: fixed this in https://github.com/wmde/DeepCat-Gadget/commit/829cd78d38f8e86fd6efee3d673d549ac8f56297 :->

**Comment by WMDE-Fisch** [comment]
> But even this code should not be here but in the backend.

You mean the JSONP Interface or CatGraph itself? 

> I found the relevant lines in core:

Thanks, I also thought that way but the obviously failed to find the part in core.

> I confused this with the zero width space (\u200B), which is actually used in category names.

Are you sure? Do you have an example?

**Comment by WMDE-Fisch** [comment]
After talking to Leszek and Adam we looked for the ZWSP in different Wikis (Japanese, Chinese, Korean ) and did not find any category/page title. It seems it really is filtered. 

For now I will adjust the filter in the Gadget to the RegExp in core.

**Comment by WMDE-Fisch** [comment]
> I confused this with the zero width space (\u200B), which is actually used in category names.

Wait now _I_ am confused. Where do you see this in the code above? ^^

**Comment by thiemowmde** [comment]
\u200B is in the original regex above, but probably shouldn't.

Is there some new code in core filtering the zero width space? As I said, I know this character is used in titles. I redid my search and found the example: In https://de.wikipedia.org/wiki/Brüder_Grimm?action=edit&oldid=111877137 the interwiki link to the bowiki contains a zero width space.

**Pull request merged**